---
title: "scDiffCom: Investigating changes in intercellular communication"
author: "Cyril Lagger, Eugen Ursu"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scDiffCom: Investigating changes in intercellular communication}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We show here how to use [scDiffCom](https://github.com/CyrilLagger/scDiffCom)
to investigate changes in intercellular communication from scRNA-seq data.

## Standard Workflow

### Setup a Seurat object

`scDiffCom` requires as input a [Seurat](https://satijalab.org/seurat/) object
that must satisfy some conditions:

* Cells should have been previously annotated with cell-types (such as
explained for example
[here](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html)).
* `meta.data` of the `Seurat` object should contain a column that
indicates to which of two groups each cell belongs to. Note that although the 
main aim of `scDiffCom` is to perform differential analysis, it is also possible
to only perform a detection analysis (see below).
* The slot `data` of the `Seurat` object should contain standard log-normalize
values.

For the purpose of this vignette, we use a dummy Seurat object available as
part of `scDiffCom`. It corresponds to the liver scRNA-seq dataset from
[Tabula Muris Senis](https://tabula-muris-senis.ds.czbiohub.org/). Note that
after normalization and pre-processing, the object has been down-sampled 
(both genes and cells) to keep it light. **As such, all results below are not
intended to convey any meaningful biology.** 

```{r setup}
library(scDiffCom)
library(Seurat)

seurat_object <- scDiffCom::seurat_sample_tms_liver
seurat_object

# cell-type annotations are stored in "cell_type"
head(seurat_object$cell_type)

# the grouping condition (here by age) is stored in "age_group"
head(seurat_object$age_group)
table(seurat_object$age_group)

```

### Database of ligand-receptor interactions

As in several other intercellular communication studies, `scDiffCom` infers
cell type to cell type communication patterns based on the gene expression of
ligands and receptors known to interact. The package provides both a human 
database and a mouse database of ligand-receptor interactions (LRIs)
that have been retrieved and processed from eight public resources.

As shown below, `scDiffCom` uses these databases internally and the user
do not need to call them explicitly to perform any interaction analysis. They
are nevertheless accessible directly from the package as follows:


```{r}

# Loading the mouse LRIs
data(LRI_mouse)

# actual data.table with the interactions
head(LRI_mouse$LRI_curated)

# GO term annotation
head(LRI_mouse$LRI_curated_GO)

# KEGG Pathway annotation
head(LRI_mouse$LRI_curated_KEGG)


```


### Run a standard differential analysis 

All differential intercellular communication results can be obtained from a single call to the function `run_interaction_analysis`. For a more advanced
usage and control of all parameters, please see the relevant section below and look at `?run_interation_analysis`.

In standard mode, it will perform the following steps:
<details>
<summary>Display the steps</summary>
 * Preprocess the data and meta.data of the Seurat object.
 * Build a table of all potential cell-cell interactions (CCIs) of the form
 (Emitter Cell Type, Receiver Cell Type, Ligand Gene(s), Receptor Gene(s)).
 They are obtained by comparing the genes in the Seurat object to those of our
 internal database of LRIs.
 * Compute scores, detection rates,... for each CCI.
 * Perform two different permutation tests to infer both the
 specificity (as originally defined by
 [CellPhoneDB](https://www.cellphonedb.org/)) and significance of change of 
 each CCI.
 * Based on their computed attributes, CCIs are either filtered out or
 assigned to a regulation category: UP, DOWN, FLAT, NSC (non-significant 
 change).
 * Perform an over-representation analysis to infer dominant communication 
 changes at the level of various categories (LRI, cell types, GO Terms,
 KEGG pathways,...).
 </details>&nbsp;
 
By default `run_interaction_analysis` performs 1000 iterations for the
permutation tests. In such as case, the analysis should only take a couple of
minutes when done sequentially on the Seurat object considered here. For a real
scenario with a bigger object and more iterations (we recommend 10'000),the analysis can be run in parallel by loading the `future` package and setting the `plan` accordingly.


```{r}

# access future options
library(future)
plan(sequential) # sequentially in the current R process, equivalent to do nothing
#plan(multisession, workers = 4) # background R sessions
#plan(multicore, workers = 4) # forked R processes, not Windows/not RStudio

# run the interaction analysis between young and old cells with
# default parameters
aging_example <- run_interaction_analysis(
  seurat_object = seurat_object,
  LRI_species = "mouse",
  seurat_celltype_id = "cell_type",
  seurat_condition_id = list(
    column_name = "age_group",
    cond1_name = "YOUNG",
    cond2_name = "OLD"
  )
)

```

### Explore the results

The output of `run_interaction_analysis` is an S4 object of class
`scDiffCom`. We recommend to access the slot of the object by using
the provided accessors.

```{r}
aging_example

# retrieving the list of parameters used to perform the analysis
head(GetParameters(aging_example), 4)

```

The main results are stored as a data.table (one row per detected CCI) in
the slot `cci_table_detected`. Each CCI is annotated with several attributes
including logFC, adjusted p-values, final regulation assignment, etc.

```{r}

CCI_detected <- GetTableCCI(aging_example, "detected", simplified = TRUE)

table(CCI_detected$REGULATION)

top10_up_regulated <- CCI_detected[REGULATION == "UP"][order(-LOGFC)][1:10]
top10_up_regulated$CCI

top10_down_regulated <- CCI_detected[REGULATION == "DOWN"][order(LOGFC)][1:10]
top10_down_regulated$CCI


```
Results from the over-representation analysis allows us to find dominant
changing patterns. This can be accessed from the slot \code{ora_table} and the
top terms can be visualized on a plot.


```{r}

# retrieve ORA results for all categories
ORA_tables <- GetTableORA(aging_example, "all", simplified = TRUE)

# investigate the up-regulated LRIs
OR_LRI_up <- ORA_tables$LRI[OR_UP > 1 & BH_P_VALUE_UP <= 0.01]$VALUE
OR_LRI_up

PlotORA(
  object = aging_example,
  category = "LRI",
  regulation = "UP"
)

# investigate the down-regulated GO-terms
OR_GO_down <- ORA_tables$GO_TERMS[OR_UP > 1 & BH_P_VALUE_UP <= 0.01]$VALUE
head(OR_GO_down)

PlotORA(
  object = aging_example,
  category = "GO_TERMS",
  regulation = "DOWN",
  GO_aspect = "biological_process"
)

```

The over-representation performed on emitter and receiver cell types allows us 
to build an interactive network of the dominant changes in the dataset.

```{r}
BuildNetwork(
  object = aging_example
)

```

## Advanced functionalities

Vignette in progress...

### Modifying filtering parameters

```{r}

aging_example_strong_logfc <- FilterCCI(
  object = aging_example,
  new_threshold_logfc = log(2)
)

```

### Performing ORA on a custom category

```{r}
library(data.table)

cell_types <-  c(
  "B cell",
  "T cell",
  "endothelial cell of hepatic sinusoid",
  "hepatocyte",
  "myeloid leukocyte"
)
cell_families <- c(
  "leukocyte",
  "leukocyte",
  "endothelial cell",
  "epithelial cell",
  "leukocyte"
)

cell_families_dt <- data.table(
  EMITTER_CELLTYPE = cell_types,
  EMITTER_CELLFAMILY = cell_families
)

RunORA(
  object = aging_example,
  categories = c("LRI"),
  extra_annotations = list(
    cell_families_dt
  ),
  overwrite = TRUE
)

```

### Running a detection analysis without differential analysis

```{r}

detection_example <- run_interaction_analysis(
  seurat_object = seurat_object,
  LRI_species = "mouse",
  seurat_celltype_id = "cell_type",
  seurat_condition_id = NULL,
  iterations = 100
)

```

### Returning null distributions
