---
title: "Introduction to scDiffCom"
description: >
  Learn how to use scDiffCom to perform differential intercellular communication
  analysis.
author: "Cyril Lagger, Eugen Ursu"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to scDiffCom}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Overview and terminology

scDiffCom infers how intercellular communication changes between two conditions
of interest from scRNA-seq data. The package can also be used without
performing differential analysis if one is not interested in comparing
conditions but only in investigating detected cell-cell interactions.

In addition to this vignette, we also recommend reading our manuscript
([doi:10.1101/2021.08.13.456238](https://www.biorxiv.org/content/10.1101/2021.08.13.456238v1))
for more information about the statistical
methods used by the package.

Important terminology:

* LRI (ligand-receptor interaction): a set of genes whose proteins are known
to interact during extracellular signalling. The package distinguishes simple
LRIs (e.g. `Apoe:Ldlr`) and complex/heteromeric LRIs (e.g.
`Col3a1:Itgb1_Itga2`).
* CCI (cell-cell interaction): a communication signal of the form
`(B cell, T cell; Apoe:Ldlr)` where `B cell` is the emitter cell type expressing
the ligand `Apoe` and `T cell` is the receiver cell type expressing the receptor
`Ldlr`.

**Note: the toy-model results below do not convey any meaningful biology.**

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE
)
```

## Standard differential analysis

### Load essential libraries

```{r setup, warning = FALSE, message = FALSE}
library(Seurat)
library(scDiffCom)
library(data.table)
```

### Prepare a Seurat object

The input of `scDiffCom` must be a [Seurat](https://satijalab.org/seurat/)
object with cells annotated by cell types. A pairwise condition on the cells is
also necessary for the differential analysis.

As a toy-model example, we use a down-sampled Seurat object containing mouse
liver cells from Tabula Muris Senis. This object is part of the package:

```{r}
seurat_object <- scDiffCom::seurat_sample_tms_liver
seurat_object

# The object already contains a column "cell_type" in its meta.data
table(seurat_object[["cell_type"]])

# Cells can be grouped based on mice age
table(seurat_object[["age_group"]])
```

### Run default analysis

The full detection and differential analysis is performed by the function
`run_interaction_analysis`. Several permutation tests are performed to assess
the significance of both the specificity and the differential expression of each
potential cell-cell interaction (CCI). 1000 permutations are done by default
as this is fast and sufficient for a preliminary exploratory analysis.
Otherwise, we recommend 10'000 permutations. Parallel computing can be easily
enabled by loading the
[future](https://cran.r-project.org/package=future) package and setting the
`plan` accordingly:

```{r, message = FALSE, warning = FALSE}
# Load future (optional)
library(future)
plan(sequential) # sequentially in the current R process, equivalent to do nothing
#plan(multisession, workers = 4) # background R sessions
#plan(multicore, workers = 4) # forked R processes, not Windows/not RStudio
```

Running the analysis only requires to specify the species, the names of 
the columns where the cell-type annotation and pairwise condition are stored
and the explicit names of the two conditions:
```{r}
# Run differential analysis with default parameters
scdiffcom_object <- run_interaction_analysis(
  seurat_object = seurat_object,
  LRI_species = "mouse",
  seurat_celltype_id = "cell_type",
  seurat_condition_id = list(
    column_name = "age_group",
    cond1_name = "YOUNG",
    cond2_name = "OLD"
  )
)
```

The output of `run_interaction_analysis` is an S4 object of class
`scDiffCom`:
```{r}
scdiffcom_object
```

### Explore the results from the Shiny app

A Shiny app can be launched from each `scDiffCom` object to interactively
visualize the detected CCIs and how they are differentially expressed. 
Over-representation results are also directly available from the app. As an
option, it is possible to first reduce over-represented GO terms by semantic
similarity such that they will be displayed as a treemap. Also note that you
might be prompted to install several packages.

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!require("shiny")) install.packages("shiny")
if (!require("shinyWidgets")) install.packages("shinyWidgets")
if (!require("shinythemes")) install.packages("shinythemes")
if (!require("DT")) install.packages("DT")
if (!require("kableExtra")) install.packages("kableExtra")
if (!require("RColorBrewer")) install.packages("RColorBrewer")
if (!require("igraph")) install.packages("igraph")
if (!require("visNetwork")) install.packages("visNetwork")
if (!require("plotly")) install.packages("plotly")
if (!require("GOSemSim")) BiocManager::install("GoSemSim")
if (!require("rrvgo")) BiocManager::install("rrvgo")
if (!require("org.Mm.eg.db")) BiocManager::install("org.Mm.eg.db")

reduced_go_terms <- ReduceGO(scdiffcom_object) # optional

BuildShiny(
  scdiffcom_object,
  reduced_go_table = reduced_go_terms # NULL by default
)
```

### Explore the results manually

Instead of using the Shiny app, the results can also be extracted directly
from the `scDiffCom` object with the provided accessors:

* `GetParameters()` returns the list of parameters used by 
`run_interaction_analysis()`
* `GetTableCCI()` returns a (complete or simplified) data.table either of
the detected CCIs or of all the hypothetical CCIs. 
* `GetTableORA()` returns a list of (complete or simplified) data.tables
with over-representation results.

From there, results can be analysed and visualised from either custom
or `scDiffCom` functions. 

```{r}
# Retrieve and display all detected CCIs
CCI_detected <- GetTableCCI(scdiffcom_object, type = "detected", simplified = TRUE)

# Number of CCIs per regulation type (here with age)
table(CCI_detected$REGULATION)

# Retrieve the ORA results
ORA_results <- GetTableORA(scdiffcom_object, categories = "all", simplified = TRUE)

# Categories available
names(ORA_results)
```

Custom volcano plot of CCIs:
```{r,out.width="100%", warning = FALSE, message = FALSE}
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)

ggplot(
  CCI_detected,
  aes(
     x = LOGFC,
     y = -log10(BH_P_VALUE_DE + 1E-2),
     colour = REGULATION
  )
) + geom_point(
) + scale_colour_manual(
  values = c("UP" = "red", "DOWN" = "blue", "FLAT" = "green", "NSC" = "grey")
) + xlab(
  "log(FC)"
) + ylab(
  "-log10(Adj. p-value)"
)
```


Dot plots of over-represented terms:
```{r, out.width="100%"}
# Plot the most over-represented up-regulated LRIs
# PlotORA returns a ggplot object that you can further optimize (e.g. here to place the legend)
PlotORA(
  object = scdiffcom_object,
  category = "LRI",
  regulation = "UP"
) + theme(
  legend.position = c(0.85, 0.4),
  legend.key.size = unit(0.4, "cm")
)
```

A network of over-represented cell types and cell-type to cell-type
interactions:
```{r, warning = FALSE, message = FALSE}
if (!require("visNetwork")) install.packages("visNetwork")
if (!require("igraph")) install.packages("igraph")
if (!require("kableExtra")) install.packages("kableExtra")
if (!require("RColorBrewer")) install.packages("RColorBrewer")

BuildNetwork(
  object = scdiffcom_object
)
```


## Detection-only analysis

If one has no condition to compare, `scDiffCom` can still be used to return
detected CCIs only:
```{r}
scdiffcom_detection_only <- run_interaction_analysis(
  seurat_object = seurat_object,
  LRI_species = "mouse",
  seurat_celltype_id = "cell_type",
  seurat_condition_id = NULL
)
```

Although it is currently not possible to explore such results with a Shiny app,
such functionality should be added in a future release of the package (
as well as with other downstream analysis tools).

## Database of ligand-receptor interactions

`scDiffcom` infers cell type to cell type communication patterns from the
expression of genes known to be involved in ligand-receptor interactions
(LRIs). The package contains it own internal databases of curated LRIs (for
human, mouse and rat), retrieved from previous studies.

As shown above, there is no need to call the LRI databases explicitly when
performing an analysis. However, they can be accessed and explored as follows:
```{r}
# Load, e.g., the mouse database
data(LRI_mouse)

# Display the data.table of LRIs (more information available in other columns)
LRI_mouse$LRI_curated[, c("LRI")]

# Display the data.table of GO Terms attached to each LRI
LRI_mouse$LRI_curated_GO

# Information about GO terms (more information available in other columns)
# LEVEL corresponds to the depth of the GO term in the GO graph
data(gene_ontology_level)
gene_ontology_level[, c("NAME", "LEVEL")]

```

## Additional functionalities

### Modifying default parameters

`run_interaction_analysis` depends on several parameters that can be modified.
We encourage to carefully read the documentation (`?run_interaction_analysis`).
For convenience, the input parameters are stored in the returned object in the
slot `parameters`:
```{r}
# display the first three parameters
head(GetParameters(scdiffcom_object), 3)
```

Among those parameters are four filtering thresholds that can be modified to
recompute the results without having to perform the time-consuming 
permutation analyses again. This works because we store an intermediate
table of all hypothetic CCIs in the slot `cci_table_raw`. The disadvantage is
that storing this data.table can make the object heavy depending on the
dataset. You can either decide to keep this table allowing you to quickly update
your results in case you want to modify some filtering parameters or to erase
this table to save space.

```{r, error=TRUE}
# create a new object with a different log fold change threshold
scdiffcom_object_stringent_logfc <- FilterCCI(
  object = scdiffcom_object,
  new_threshold_logfc = log(2)
)

# create a copy of the original object, but without the heavy intermediate table
scdiffcom_object_light <- EraseRawCCI(scdiffcom_object)

# the light object cannot be updated anymore
FilterCCI(
  object = scdiffcom_object_light,
  new_threshold_logfc = log(2)
)
```

### Performing ORA on custom categories

By default, ORA is performed on a set of categories that are always present for
any dataset analysed (LRI, GO Terms, etc). However, you might want to find
over-represented results for a category only relevant to your dataset. This is 
possible to achieve as long as you are able to build a relationship between
the terms of your new category and those of a default category. As an example,
we show here how we can perform ORA on cell-type families:

```{r}
# Create a data.table that associates cell-type families to cell types
# Here we use the default category EMITTER_CELLTYPE (RECEIVER_CELLTYPE would also work)
cell_families_relation <- data.table(
  EMITTER_CELLTYPE = c(
    "B cell",
    "T cell",
    "endothelial cell of hepatic sinusoid",
    "hepatocyte",
    "myeloid leukocyte"
  ),
  EMITTER_CELLFAMILY = c(
    "leukocyte",
    "leukocyte",
    "endothelial cell",
    "epithelial cell",
    "leukocyte"
  )
)

# Run ORA with the cell type families as extra annotation
scdiffcom_object <- RunORA(
  object = scdiffcom_object,
  extra_annotations = list(
    cell_families_relation
  ),
  overwrite = FALSE
)
```

The object now contains a new table in its slot `ora_table`.
