<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scDiffCom: Investigating changes in intercellular communication â€¢ scDiffCom</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scDiffCom: Investigating changes in intercellular communication">
<meta property="og:description" content="scDiffCom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiffCom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scDiffCom-vignette.html">scDiffCom: Investigating changes in intercellular communication</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scDiffCom-vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="scDiffCom-vignette_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="scDiffCom-vignette_files/vis-4.20.1/vis.css" rel="stylesheet">
<script src="scDiffCom-vignette_files/vis-4.20.1/vis.min.js"></script><script src="scDiffCom-vignette_files/visNetwork-binding-2.0.9/visNetwork.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scDiffCom: Investigating changes in intercellular communication</h1>
                        <h4 class="author">Cyril Lagger, Eugen Ursu</h4>
            
      
      
      <div class="hidden name"><code>scDiffCom-vignette.Rmd</code></div>

    </div>

    
    
<p>We show here how to use <a href="https://github.com/CyrilLagger/scDiffCom">scDiffCom</a> to investigate changes in intercellular communication from scRNA-seq data.</p>
<div id="standard-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#standard-workflow" class="anchor"></a>Standard Workflow</h2>
<div id="setup-a-seurat-object" class="section level3">
<h3 class="hasAnchor">
<a href="#setup-a-seurat-object" class="anchor"></a>Setup a Seurat object</h3>
<p><code>scDiffCom</code> requires as input a <a href="https://satijalab.org/seurat/">Seurat</a> object that must satisfy some conditions:</p>
<ul>
<li>Cells should have been previoulsy annotated with cell-types (such as explained for example <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">here</a>).</li>
<li>The <code>meta.data</code> of the <code>Seurat</code> object should contain a column that indicates to which of two groups each cell belongs to. Note that although the main aim of <code>scDiffCom</code> is to perform differential analysis, it is also possible to only perform a detection analysis (see below).</li>
<li>The slot <code>data</code> of the <code>Seurat</code> object should contain standard log-normalize values.</li>
</ul>
<p>For the purpose of this vignette, we use a dummy Seurat object available as part of <code>scDiffCom</code>. It corresponds to the liver scRNA-seq dataset from the <a href="https://tabula-muris-senis.ds.czbiohub.org/">Tabula Muris Senis</a>. Note that after normalization and pre-processing, the object has been down-sampled (in both genes and cells) to keep it light. <strong>As such, all results below are not intended to convey any meaningful biology.</strong> For a real biological analysis using <code>scDiffCom</code> please see our scAgeCom analysis (article in prep.)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scDiffCom</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Attaching SeuratObject</span>

<span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu">scDiffCom</span><span class="fu">::</span><span class="va"><a href="../reference/seurat_sample_tms_liver.html">seurat_sample_tms_liver</a></span>
<span class="va">seurat_object</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 726 features across 468 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (726 features, 0 variable features)</span>

<span class="co"># cell-type annotations are stored in "cell_type"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span>
<span class="co">#&gt;  A16_B000767_B009945_S16.mm10-plus-4-0  A16_D045313_B009942_S16.mm10-plus-4-0 </span>
<span class="co">#&gt;                               "T cell" "endothelial cell of hepatic sinusoid" </span>
<span class="co">#&gt;  A21_B001247_B009941_S21.mm10-plus-4-0  A22_B001247_B009941_S22.mm10-plus-4-0 </span>
<span class="co">#&gt;                               "B cell"                               "B cell" </span>
<span class="co">#&gt;    A2_D045313_B009942_S2.mm10-plus-4-0    A5_B001247_B009941_S5.mm10-plus-4-0 </span>
<span class="co">#&gt;                           "hepatocyte"                               "T cell"</span>

<span class="co"># the grouping condition (here by age) is stored in "age_group"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">$</span><span class="va">age_group</span><span class="op">)</span>
<span class="co">#&gt; A16_B000767_B009945_S16.mm10-plus-4-0 A16_D045313_B009942_S16.mm10-plus-4-0 </span>
<span class="co">#&gt;                                 "OLD"                                 "OLD" </span>
<span class="co">#&gt; A21_B001247_B009941_S21.mm10-plus-4-0 A22_B001247_B009941_S22.mm10-plus-4-0 </span>
<span class="co">#&gt;                                 "OLD"                                 "OLD" </span>
<span class="co">#&gt;   A2_D045313_B009942_S2.mm10-plus-4-0   A5_B001247_B009941_S5.mm10-plus-4-0 </span>
<span class="co">#&gt;                                 "OLD"                                 "OLD"</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">$</span><span class="va">age_group</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   OLD YOUNG </span>
<span class="co">#&gt;   250   218</span></code></pre></div>
</div>
<div id="database-of-ligand-receptor-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#database-of-ligand-receptor-interactions" class="anchor"></a>Database of ligand-receptor interactions</h3>
<p>As in several other intercellular communication studies, <code>scDiffCom</code> infers cell type to cell type communication patterns based on the gene expression of ligands and receptors known to interact. The package provides both a human database and a mouse database of ligand-receptor interactions (LRIs) that have been retrieved and processed from eight public resources.</p>
<p>As shown below, <code>scDiffCom</code> uses these databases internally and the user do not need to call them explicitly to perform any interaction analysis. They are nevertheless accessible directly from the package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Loading the mouse LRIs</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">LRI_mouse</span><span class="op">)</span>

<span class="co"># actual data.table with the interactions</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;                     LRI      LIGAND_1 LIGAND_2 RECEPTOR_1 RECEPTOR_2 RECEPTOR_3</span>
<span class="co">#&gt; 1: 1700013F07Rik:Plscr4 1700013F07Rik     &lt;NA&gt;     Plscr4       &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt; 2:   2300002M23Rik:Ddr1 2300002M23Rik     &lt;NA&gt;       Ddr1       &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt; 3:  9530003J23Rik:Itgal 9530003J23Rik     &lt;NA&gt;      Itgal       &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt; 4:             A2m:Lrp1           A2m     &lt;NA&gt;       Lrp1       &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt; 5:         Aanat:Mtnr1a         Aanat     &lt;NA&gt;     Mtnr1a       &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt; 6:         Aanat:Mtnr1b         Aanat     &lt;NA&gt;     Mtnr1b       &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;                                                DATABASE</span>
<span class="co">#&gt; 1:                           scTensor;SingleCellSignalR</span>
<span class="co">#&gt; 2:                           scTensor;SingleCellSignalR</span>
<span class="co">#&gt; 3:                           scTensor;SingleCellSignalR</span>
<span class="co">#&gt; 4: connectomeDB2020;NicheNet;scTensor;SingleCellSignalR</span>
<span class="co">#&gt; 5: connectomeDB2020;NicheNet;scTensor;SingleCellSignalR</span>
<span class="co">#&gt; 6: connectomeDB2020;NicheNet;scTensor;SingleCellSignalR</span>

<span class="co"># GO term annotation</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated_GO</span><span class="op">)</span>
<span class="co">#&gt;                     LRI      GO_ID            GO_NAME             ASPECT LEVEL</span>
<span class="co">#&gt; 1: 1700013F07Rik:Plscr4 GO:0005575 cellular_component cellular_component     1</span>
<span class="co">#&gt; 2: 1700013F07Rik:Plscr4 GO:0044464          cell part cellular_component     2</span>
<span class="co">#&gt; 3: 1700013F07Rik:Plscr4 GO:0003674 molecular_function molecular_function     1</span>
<span class="co">#&gt; 4: 1700013F07Rik:Plscr4 GO:0005488            binding molecular_function     2</span>
<span class="co">#&gt; 5: 1700013F07Rik:Plscr4 GO:0008150 biological_process biological_process     1</span>
<span class="co">#&gt; 6:   2300002M23Rik:Ddr1 GO:0003674 molecular_function molecular_function     1</span>

<span class="co"># KEGG Pathway annotation</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated_KEGG</span><span class="op">)</span>
<span class="co">#&gt;            LRI       KEGG_ID                    KEGG_NAME</span>
<span class="co">#&gt; 1: Apoa1:Abca1 path:mmu04975 Fat digestion and absorption</span>
<span class="co">#&gt; 2: Apoa1:Abca1 path:mmu04979       Cholesterol metabolism</span>
<span class="co">#&gt; 3: Apoa1:Abca1 path:mmu05417    Lipid and atherosclerosis</span>
<span class="co">#&gt; 4:  Apoe:Abca1 path:mmu04979       Cholesterol metabolism</span>
<span class="co">#&gt; 5: Calm1:Abca1 path:mmu05417    Lipid and atherosclerosis</span>
<span class="co">#&gt; 6: Calm2:Abca1 path:mmu05417    Lipid and atherosclerosis</span></code></pre></div>
</div>
<div id="run-a-standard-differential-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#run-a-standard-differential-analysis" class="anchor"></a>Run a standard differential analysis</h3>
<p>All differential intercellular communication results can be obtained from a single call to the function <code>run_interaction_analysis</code>. For a more advanced usage and control of all parameters, please see the relevant section below and look at <code>?run_interation_analysis</code>.</p>
<p>In standard mode, it will perform the following steps:</p>
<ul>
<li>Preprocess the data and meta.data of the Seurat object.</li>
<li>Build a table of all potential cell-cell interactions (CCIs) of the form (Emitter Cell Type, Receiver Cell Type, Ligand Gene(s), Receptor Gene(s)). They are obtained by comparing the genes in the Seurat object to those of our internal database of LRIs.</li>
<li>Compute scores, detection rates,â€¦ for each CCI.</li>
<li>Perform two different permutation tests to infer both the specificity (as originally defined by <a href="https://www.cellphonedb.org/">CellPhoneDB</a>) and significance of change of each CCI.</li>
<li>Based on their computed attributes, CCIs are either filtered out or assigned to a regulation category: UP, DOWN, FLAT, NSC (non-significant change).</li>
<li>Perform an over-representation analysis to infer dominant communication changes at the level of various categories (LRI, cell types, GO Terms, KEGG pathways,â€¦).</li>
</ul>
<p>By default <code>run_interaction_analysis</code> performs 1000 iterations for the permutation tests. In such as case, the analysis should only take a couple of minutes when run sequentially on the Seurat object considered here. For a real scenario with a bigger object and potentially more iterations (e.g.Â 10â€™0000), we recommend to run the analysis in parallel by loading the <code>future</code> package and setting the <code>plan</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># access future options</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/future">future</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/future/man/plan.html">plan</a></span><span class="op">(</span><span class="va">sequential</span><span class="op">)</span> <span class="co"># sequentially and in the current R process</span>
<span class="co">#plan(multisession, workers = 4) # background R sessions</span>
<span class="co">#plan(multicore, workers = 4) # forked R processes, not Windows/not RStudio</span>

<span class="co"># run the interaction analysis between young and old cells with</span>
<span class="co"># default parameters</span>
<span class="va">aging_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_interaction_analysis.html">run_interaction_analysis</a></span><span class="op">(</span>
  seurat_object <span class="op">=</span> <span class="va">seurat_object</span>,
  LRI_species <span class="op">=</span> <span class="st">"mouse"</span>,
  seurat_celltype_id <span class="op">=</span> <span class="st">"cell_type"</span>,
  seurat_condition_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    column_name <span class="op">=</span> <span class="st">"age_group"</span>,
    cond1_name <span class="op">=</span> <span class="st">"YOUNG"</span>,
    cond2_name <span class="op">=</span> <span class="st">"OLD"</span>
  <span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt; Extracting data from assay 'RNA' and slot 'data' (assuming normalized log1p-transformed data).</span>
<span class="co">#&gt; Converting normalized data from log1p-transformed to non-log1p-transformed.</span>
<span class="co">#&gt; Input data: 726 genes, 468 cells and 5 cell-types.</span>
<span class="co">#&gt; Input ligand-receptor database: 4990 mouse interactions.</span>
<span class="co">#&gt; Number of LRIs that match to genes present in the dataset: 1331.</span>
<span class="co">#&gt; Type of analysis to be performed: differential analysis between YOUNG and OLD cells.</span>
<span class="co">#&gt; Total number of potential cell-cell interactions (CCIs): 33275 (5 * 5 * 1331).</span>
<span class="co">#&gt; Performing permutation analysis (1000 iterations by batches of 1000) on 10957 potential CCIs.</span>
<span class="co">#&gt; Performing batch 1 of 1.</span>
<span class="co">#&gt; Filtering and cleaning 'raw' CCIs.</span>
<span class="co">#&gt; Returning 2305 detected CCIs.</span>
<span class="co">#&gt; Performing over-representation analysis on the categories: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span>
<span class="co">#&gt; Successfully returning final scDiffCom object.</span></code></pre></div>
</div>
<div id="explore-the-results" class="section level3">
<h3 class="hasAnchor">
<a href="#explore-the-results" class="anchor"></a>Explore the results</h3>
<p>The output of <code>run_interaction_analysis</code> is an S4 object of class <code>scDiffCom</code>. We recommend to access the slot of the object by using the provided accessors</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aging_example</span>
<span class="co">#&gt; An object of class scDiffCom with name scDiffCom_object</span>
<span class="co">#&gt; Analysis performed: differential analysis between YOUNG and OLD cells</span>
<span class="co">#&gt; 2305 detected CCIs across 5 cell types</span>
<span class="co">#&gt; Over-representation results for LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS</span>

<span class="co"># retrieving the list of parameters used to perform the analysis</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/GetParameters.html">GetParameters</a></span><span class="op">(</span><span class="va">aging_example</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span>
<span class="co">#&gt; $object_name</span>
<span class="co">#&gt; [1] "scDiffCom_object"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $LRI_species</span>
<span class="co">#&gt; [1] "mouse"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $seurat_celltype_id</span>
<span class="co">#&gt; [1] "cell_type"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $seurat_condition_id</span>
<span class="co">#&gt; $seurat_condition_id$column_name</span>
<span class="co">#&gt; [1] "age_group"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $seurat_condition_id$cond1_name</span>
<span class="co">#&gt; [1] "YOUNG"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $seurat_condition_id$cond2_name</span>
<span class="co">#&gt; [1] "OLD"</span></code></pre></div>
<p>The main results are stored as a data.table (one row per detected CCI) in the slot <code>cci_table_detected</code>. Each CCI is annotated with several attributes including logFC, adjusted p-values, final regulation assignment, etc.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">CCI_detected</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetTableCCI.html">GetTableCCI</a></span><span class="op">(</span><span class="va">aging_example</span>, <span class="st">"detected"</span>, simplified <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">CCI_detected</span><span class="op">$</span><span class="va">REGULATION</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; DOWN FLAT  NSC   UP </span>
<span class="co">#&gt;  132 1143  994   36</span>

<span class="va">top10_up_regulated</span> <span class="op">&lt;-</span> <span class="va">CCI_detected</span><span class="op">[</span><span class="va">REGULATION</span> <span class="op">==</span> <span class="st">"UP"</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">LOGFC</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="va">top10_up_regulated</span><span class="op">$</span><span class="va">CCI</span>
<span class="co">#&gt;  [1] "myeloid leukocyte_myeloid leukocyte_Alox5:Alox5ap"       </span>
<span class="co">#&gt;  [2] "B cell_T cell_H2-Q6:Cd8b1"                               </span>
<span class="co">#&gt;  [3] "myeloid leukocyte_T cell_Crlf2:Il7r_Crlf2"               </span>
<span class="co">#&gt;  [4] "endothelial cell of hepatic sinusoid_T cell_H2-T22:Cd8b1"</span>
<span class="co">#&gt;  [5] "B cell_T cell_H2-M3:Cd8b1"                               </span>
<span class="co">#&gt;  [6] "B cell_T cell_H2-D1:Cd8b1"                               </span>
<span class="co">#&gt;  [7] "B cell_T cell_H2-K1:Cd8b1"                               </span>
<span class="co">#&gt;  [8] "hepatocyte_T cell_H2-Q10:Cd8b1"                          </span>
<span class="co">#&gt;  [9] "B cell_T cell_H2-Q6:Cd8a"                                </span>
<span class="co">#&gt; [10] "T cell_myeloid leukocyte_Ltb:Tnfrsf1a"</span>

<span class="va">top10_down_regulated</span> <span class="op">&lt;-</span> <span class="va">CCI_detected</span><span class="op">[</span><span class="va">REGULATION</span> <span class="op">==</span> <span class="st">"DOWN"</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">LOGFC</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="va">top10_down_regulated</span><span class="op">$</span><span class="va">CCI</span>
<span class="co">#&gt;  [1] "B cell_endothelial cell of hepatic sinusoid_Gcg:Ramp2"        </span>
<span class="co">#&gt;  [2] "B cell_endothelial cell of hepatic sinusoid_Iapp:Calcrl"      </span>
<span class="co">#&gt;  [3] "B cell_endothelial cell of hepatic sinusoid_Iapp:Ramp2"       </span>
<span class="co">#&gt;  [4] "B cell_endothelial cell of hepatic sinusoid_Ins2:Insr"        </span>
<span class="co">#&gt;  [5] "B cell_endothelial cell of hepatic sinusoid_Pyy:Dpp4"         </span>
<span class="co">#&gt;  [6] "B cell_hepatocyte_Iapp:Ide"                                   </span>
<span class="co">#&gt;  [7] "B cell_hepatocyte_Ins2:Ide"                                   </span>
<span class="co">#&gt;  [8] "B cell_myeloid leukocyte_Ptpn6:Fcrls"                         </span>
<span class="co">#&gt;  [9] "T cell_T cell_Lamc1:Itgb1_Itga2"                              </span>
<span class="co">#&gt; [10] "T cell_endothelial cell of hepatic sinusoid_Lamc1:Itga1_Itgb1"</span></code></pre></div>
<p>Results from the over-representation analysis allows us to find dominant changing patterns. This can be accessed from the slot  and the top terms can be visualized on a plot.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># retrieve ORA results for all categories</span>
<span class="va">ORA_tables</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetTableORA.html">GetTableORA</a></span><span class="op">(</span><span class="va">aging_example</span>, <span class="st">"all"</span>, simplified <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># investigate the up-regulated LRIs</span>
<span class="va">OR_LRI_up</span> <span class="op">&lt;-</span> <span class="va">ORA_tables</span><span class="op">$</span><span class="va">LRI</span><span class="op">[</span><span class="va">OR_UP</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">BH_P_VALUE_UP</span> <span class="op">&lt;=</span> <span class="fl">0.01</span><span class="op">]</span><span class="op">$</span><span class="va">VALUE</span>
<span class="va">OR_LRI_up</span>
<span class="co">#&gt;  [1] "B2m:Cd3g"      "H2-D1:Cd8a"    "H2-D1:Cd8b1"   "H2-K1:Cd8a"   </span>
<span class="co">#&gt;  [5] "H2-K1:Cd8b1"   "H2-Q6:Cd8b1"   "H2-T22:Cd8b1"  "H2-T23:Cd8b1" </span>
<span class="co">#&gt;  [9] "Lgals1:Lgals1" "Ltb:Tnfrsf1a"</span>

<span class="fu"><a href="../reference/PlotORA.html">PlotORA</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">aging_example</span>,
  category <span class="op">=</span> <span class="st">"LRI"</span>,
  regulation <span class="op">=</span> <span class="st">"UP"</span>
<span class="op">)</span></code></pre></div>
<p><img src="scDiffCom-vignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># investigate the down-regulated GO-terms</span>
<span class="va">OR_GO_down</span> <span class="op">&lt;-</span> <span class="va">ORA_tables</span><span class="op">$</span><span class="va">GO_TERMS</span><span class="op">[</span><span class="va">OR_UP</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">BH_P_VALUE_UP</span> <span class="op">&lt;=</span> <span class="fl">0.01</span><span class="op">]</span><span class="op">$</span><span class="va">VALUE</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">OR_GO_down</span><span class="op">)</span>
<span class="co">#&gt; [1] "cellular glucose homeostasis"                             </span>
<span class="co">#&gt; [2] "B cell activation involved in immune response"            </span>
<span class="co">#&gt; [3] "mature B cell differentiation involved in immune response"</span>
<span class="co">#&gt; [4] "plasma cell differentiation"                              </span>
<span class="co">#&gt; [5] "mature B cell differentiation"                            </span>
<span class="co">#&gt; [6] "immune system process"</span>

<span class="fu"><a href="../reference/PlotORA.html">PlotORA</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">aging_example</span>,
  category <span class="op">=</span> <span class="st">"GO_TERMS"</span>,
  regulation <span class="op">=</span> <span class="st">"DOWN"</span>,
  GO_aspect <span class="op">=</span> <span class="st">"biological_process"</span>
<span class="op">)</span></code></pre></div>
<p><img src="scDiffCom-vignette_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<p>The over-representation performed on emitter and receiver cell types allows us to build an interactive network of the dominant changes in the dataset.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/BuildNetwork.html">BuildNetwork</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">aging_example</span>,
  network_type <span class="op">=</span> <span class="st">"ORA_network"</span>,
  layout_type <span class="op">=</span> <span class="st">"bipartite"</span>
<span class="op">)</span></code></pre></div>
<div id="htmlwidget-409849e39179307f2e17" style="width:100%;height:432.632880098888px;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-409849e39179307f2e17">{"x":{"nodes":{"name":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","myeloid leukocyte (E)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"vertex_types":[true,true,true,true,true,false,false,false,false,false],"num_cells":[42,42,50,50,50,42,42,50,50,50],"vertex.size":[5,5,5,5,5,5,5,5,5,5],"id":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","myeloid leukocyte (E)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"label":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","myeloid leukocyte (E)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"value":[5,5,5,5,5,5,5,5,5,5],"color.background":["#f8961e","#f8961e","#f8961e","#f8961e","#f8961e","#f8961e","#f8961e","#f8961e","#f8961e","#f8961e"],"color.border":["#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590"],"color.highlight.background":["#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B"],"color.highlight.border":["#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590"],"color.hover.background":["#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B","#43AA8B"],"color.hover.border":["#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590"],"shadow":[true,true,true,true,true,true,true,true,true,true],"title":["<h3> B cell (E) <\/h3><p> Avg. Number of Cells: 42 <\/p>","<h3> T cell (E) <\/h3><p> Avg. Number of Cells: 42 <\/p>","<h3> endothelial cell of hepatic sinusoid (E) <\/h3><p> Avg. Number of Cells: 50 <\/p>","<h3> hepatocyte (E) <\/h3><p> Avg. Number of Cells: 50 <\/p>","<h3> myeloid leukocyte (E) <\/h3><p> Avg. Number of Cells: 50 <\/p>","<h3> B cell (R) <\/h3><p> Avg. Number of Cells: 42 <\/p>","<h3> T cell (R) <\/h3><p> Avg. Number of Cells: 42 <\/p>","<h3> endothelial cell of hepatic sinusoid (R) <\/h3><p> Avg. Number of Cells: 50 <\/p>","<h3> hepatocyte (R) <\/h3><p> Avg. Number of Cells: 50 <\/p>","<h3> myeloid leukocyte (R) <\/h3><p> Avg. Number of Cells: 50 <\/p>"],"group":[true,true,true,true,true,false,false,false,false,false],"level":[1,1,1,1,1,2,2,2,2,2],"vertex_order":[1,2,4,5,3,6,7,9,10,8],"x":[-1,-1,-1,-1,-1,1,1,1,1,1],"y":[-1,-0.5,0.5,1,0,-1,-0.5,0.5,1,0]},"edges":{"from":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","hepatocyte (E)","hepatocyte (E)","hepatocyte (E)","hepatocyte (E)"],"to":["T cell (R)","T cell (R)","T cell (R)","hepatocyte (R)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"NUM_LRIS_TOTAL":[72,92,128,71,55,67,125,79,154],"NUM_LRIS_YOUNG":[53,74,117,68,49,65,122,79,152],"NUM_LRIS_OLD":[68,82,104,66,49,46,107,72,151],"NUM_LRIS_REL_DIFF":[15,8,-13,-2,0,-19,-15,-7,-1],"NUM_LRIS_UP":[10,9,5,null,null,2,null,null,null],"NUM_LRIS_DOWN":[3,6,15,10,1,13,18,5,2],"NUM_LRIS_FLAT":[24,38,46,41,39,23,55,54,97],"NUM_LRIS_NSC":[35,39,62,20,15,29,52,20,55],"NUM_LRIS_DIFF":[13,15,20,null,null,15,null,null,null],"ORA_TYPE":["UP","UP","DOWN","DOWN","FLAT","DOWN","DOWN","FLAT","FLAT"],"OR_UP":[13.6,8.76,2.81,0,0,1.99,0,0,0],"OR_DOWN":[0.709,1.16,2.34,2.84,0.3,4.28,3.05,1.12,0.205],"OR_FLAT":[0.498,0.706,0.552,1.4,2.53,0.522,0.789,2.25,1.8],"BH_P_VALUE_UP":[5.61e-07,3.02e-05,0.104,null,null,0.507,null,null,null],"BH_P_VALUE_DOWN":[0.911,0.91,0.0238,0.0238,0.608,0.00161,0.00202,0.911,0.0348],"BH_P_VALUE_FLAT":[0.0235,0.245,0.0118,0.356,0.0118,0.0457,0.386,0.01,0.01],"color.color":["#F94144","#F94144","#277DA1","#277DA1","#90BE6D","#277DA1","#277DA1","#90BE6D","#90BE6D"],"main_title":["Over-represented cell type pairs","Over-represented cell type pairs","Over-represented cell type pairs","Over-represented cell type pairs","Over-represented cell type pairs","Over-represented cell type pairs","Over-represented cell type pairs","Over-represented cell type pairs","Over-represented cell type pairs"],"width":[2.55360423994558,3.43821809359439,4.81508947694223,2.50632522693082,1.69881284537628,2.31379459434091,4.7084538336495,2.87575568968245,5.69137473758494],"color.highlight":["#F94144","#F94144","#277DA1","#277DA1","#90BE6D","#277DA1","#277DA1","#90BE6D","#90BE6D"],"color.hover":["#F94144","#F94144","#277DA1","#277DA1","#90BE6D","#277DA1","#277DA1","#90BE6D","#90BE6D"],"smooth":[true,true,true,true,true,true,true,true,true],"arrow.size":[0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5],"title":["<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio UP: <\/td>\n   <td style=\"text-align:left;\"> 13.6 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value UP: <\/td>\n   <td style=\"text-align:left;\"> 5.61e-07 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 72 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 10 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 3 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 24 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio UP: <\/td>\n   <td style=\"text-align:left;\"> 8.76 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value UP: <\/td>\n   <td style=\"text-align:left;\"> 3.02e-05 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 92 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 9 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 6 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 38 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 2.34 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.0238 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 128 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 5 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 15 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 46 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 2.84 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.0238 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 71 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> NA <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 10 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 41 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 2.53 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.0118 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 55 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> NA <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 1 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 39 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 4.28 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.00161 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 67 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 2 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 13 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 23 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 3.05 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.00202 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 125 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> NA <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 18 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 55 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 2.25 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.01 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 79 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> NA <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 5 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 54 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 1.8 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.01 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of interactions: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 154 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> NA <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 2 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 97 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","font":{"size":18,"align":"left"},"physics":false},"manipulation":{"enabled":true},"edges":{"arrows":"middle","shadow":true,"smooth":false},"physics":{"maxVelocity":10,"timestep":0.5,"enabled":false,"repulsion":{"damping":0.5}},"interaction":{"hover":true,"keyboard":{"enabled":true},"multiselect":true,"navigationButtons":false},"autoResize":true},"groups":["TRUE","FALSE"],"width":"100%","height":null,"idselection":{"enabled":false,"style":"width: 150px; height: 26px","useLabels":true,"main":"Select by id"},"byselection":{"enabled":true,"style":"width: 200px; height: 26px;\n          background: #f8f8f8;\n          color: darkblue;\n          border:none;\n          outline:none;","multiple":true,"hideColor":"rgba(200,200,200,0.5)","highlight":false,"variable":"name","main":"Select by name","values":["B cell (E)","B cell (R)","endothelial cell of hepatic sinusoid (E)","endothelial cell of hepatic sinusoid (R)","hepatocyte (E)","hepatocyte (R)","myeloid leukocyte (E)","myeloid leukocyte (R)","T cell (E)","T cell (R)"]},"main":{"text":"Over-represented cell type pairs","style":"font-family:Georgia, Times New Roman, Times, serif;font-weight:bold;font-size:20px;text-align:center;"},"submain":null,"footer":null,"background":"#F5F5F5","igraphlayout":{"type":"square"},"legend":{"width":0.2,"useGroups":false,"position":"right","ncol":1,"stepX":100,"stepY":100,"zoom":false,"edges":{"color":["#277DA1","#F94144","#90BE6D","#F9C74F"],"label":["DOWN","UP","FLAT","UP&DOWN"],"arrows":["to","to","to","to"]},"edgesToDataframe":true,"nodes":{"label":["Cell Type"],"shape":["dot"],"color":["#f8961e"]},"nodesToDataframe":true,"main":{"text":"Legend","style":"font-family:Georgia, Times New Roman, Times, serif;font-weight:bold;font-size:14px;text-align:center;"}},"datacss":"table.legend_table {\n  font-size: 11px;\n  border-width:1px;\n  border-color:#d3d3d3;\n  border-style:solid;\n}\ntable.legend_table td {\n  border-width:1px;\n  border-color:#d3d3d3;\n  border-style:solid;\n  padding: 2px;\n}\ndiv.table_content {\n  width:80px;\n  text-align:center;\n}\ndiv.table_description {\n  width:100px;\n}\n\n#operation {\n  font-size:28px;\n}\n#network-popUp {\n  display:none;\n  z-index:299;\n  width:250px;\n  height:150px;\n  background-color: #f9f9f9;\n  border-style:solid;\n  border-width:3px;\n  border-color: #5394ed;\n  padding:10px;\n  text-align: center;\n  position:fixed;\n  top:50%;  \n  left:50%;  \n  margin:-100px 0 0 -100px;  \n\n}","highlight":{"enabled":true,"hoverNearest":true,"degree":{"from":1,"to":1},"algorithm":"hierarchical","hideColor":"rgba(200,200,200,0.5)","labelOnly":true},"collapse":{"enabled":true,"fit":false,"resetHighlight":true,"clusterOptions":null,"keepCoord":true,"labelSuffix":"(cluster)"},"tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div id="advanced-functionalities" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-functionalities" class="anchor"></a>Advanced functionalities</h2>
<div id="modifying-filtering-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#modifying-filtering-parameters" class="anchor"></a>Modifying filtering parameters</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">aging_example_strong_logfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterCCI.html">FilterCCI</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">aging_example</span>,
  new_threshold_logfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt; Filtering and cleaning 'raw' CCIs.</span>
<span class="co">#&gt; Returning 2305 detected CCIs.</span>
<span class="co">#&gt; Performing over-representation analysis on the categories: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span>
<span class="co">#&gt; Erasing all previous ORA results: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span></code></pre></div>
</div>
<div id="performing-ora-on-a-custom-category" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-ora-on-a-custom-category" class="anchor"></a>Performing ORA on a custom category</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span>

<span class="va">cell_types</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">"B cell"</span>,
  <span class="st">"T cell"</span>,
  <span class="st">"endothelial cell of hepatic sinusoid"</span>,
  <span class="st">"hepatocyte"</span>,
  <span class="st">"myeloid leukocyte"</span>
<span class="op">)</span>
<span class="va">cell_families</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>
  <span class="st">"leukocyte"</span>,
  <span class="st">"leukocyte"</span>,
  <span class="st">"endothelial cell"</span>,
  <span class="st">"epithelial cell"</span>,
  <span class="st">"leukocyte"</span>
<span class="op">)</span>

<span class="va">cell_families_dt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span><span class="op">(</span>
  EMITTER_CELLTYPE <span class="op">=</span> <span class="va">cell_types</span>,
  EMITTER_CELLFAMILY <span class="op">=</span> <span class="va">cell_families</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/RunORA.html">RunORA</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">aging_example</span>,
  categories <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"LRI"</span><span class="op">)</span>,
  extra_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="va">cell_families_dt</span>
  <span class="op">)</span>,
  overwrite <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="co">#&gt; Performing over-representation analysis on the categories: LRI, EMITTER_CELLFAMILY.</span>
<span class="co">#&gt; Erasing all previous ORA results: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span>
<span class="co">#&gt; An object of class scDiffCom with name scDiffCom_object</span>
<span class="co">#&gt; Analysis performed: differential analysis between YOUNG and OLD cells</span>
<span class="co">#&gt; 2305 detected CCIs across 5 cell types</span>
<span class="co">#&gt; Over-representation results for LRI, EMITTER_CELLFAMILY</span></code></pre></div>
</div>
<div id="running-a-detection-analysis-without-differential-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#running-a-detection-analysis-without-differential-analysis" class="anchor"></a>Running a detection analysis without differential analysis</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">detection_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_interaction_analysis.html">run_interaction_analysis</a></span><span class="op">(</span>
  seurat_object <span class="op">=</span> <span class="va">seurat_object</span>,
  LRI_species <span class="op">=</span> <span class="st">"mouse"</span>,
  seurat_celltype_id <span class="op">=</span> <span class="st">"cell_type"</span>,
  seurat_condition_id <span class="op">=</span> <span class="cn">NULL</span>,
  iterations <span class="op">=</span> <span class="fl">100</span>
<span class="op">)</span>
<span class="co">#&gt; Extracting data from assay 'RNA' and slot 'data' (assuming normalized log1p-transformed data).</span>
<span class="co">#&gt; Converting normalized data from log1p-transformed to non-log1p-transformed.</span>
<span class="co">#&gt; Input data: 726 genes, 468 cells and 5 cell-types.</span>
<span class="co">#&gt; Input ligand-receptor database: 4990 mouse interactions.</span>
<span class="co">#&gt; Number of LRIs that match to genes present in the dataset: 1331.</span>
<span class="co">#&gt; Type of analysis to be performed: detection analysis without conditions.</span>
<span class="co">#&gt; Total number of potential cell-cell interactions (CCIs): 33275 (5 * 5 * 1331).</span>
<span class="co">#&gt; Performing permutation analysis (100 iterations by batches of 1000) on 8950 potential CCIs.</span>
<span class="co">#&gt; Performing batch 1 of 1.</span>
<span class="co">#&gt; Filtering and cleaning 'raw' CCIs.</span>
<span class="co">#&gt; Returning 8950 detected CCIs.</span>
<span class="co">#&gt; No over-representation analysis available for the selected parameters.</span>
<span class="co">#&gt; Successfully returning final scDiffCom object.</span></code></pre></div>
</div>
<div id="returning-null-distributions" class="section level3">
<h3 class="hasAnchor">
<a href="#returning-null-distributions" class="anchor"></a>Returning null distributions</h3>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Cyril Lagger, Eugen Ursu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
