<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Learn how to use scDiffCom to perform differential intercellular communication analysis.
">
<title>Introduction to scDiffCom • scDiffCom</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to scDiffCom">
<meta property="og:description" content="Learn how to use scDiffCom to perform differential intercellular communication analysis.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scDiffCom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/index.html">Articles</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="scDiffCom-vignette_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="scDiffCom-vignette_files/vis-9.1.0/vis-network.min.css" rel="stylesheet">
<script src="scDiffCom-vignette_files/vis-9.1.0/vis-network.min.js"></script><script src="scDiffCom-vignette_files/visNetwork-binding-2.1.2/visNetwork.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to scDiffCom</h1>
                        <h4 data-toc-skip class="author">Cyril Lagger,
Eugen Ursu</h4>
            
      
      
      <div class="d-none name"><code>scDiffCom-vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview-and-terminology">Overview and terminology<a class="anchor" aria-label="anchor" href="#overview-and-terminology"></a>
</h2>
<p>scDiffCom infers how intercellular communication changes between two
conditions of interest from scRNA-seq data. The package can also be used
without performing differential analysis if one is not interested in
comparing conditions but only in investigating detected cell-cell
interactions.</p>
<p>In addition to this vignette, we also recommend reading our (<a href="https://www.nature.com/articles/s43587-023-00514-x" class="external-link">Nature Aging
paper</a>) for more information about the statistical methods used by
the package.</p>
<p>Important terminology:</p>
<ul>
<li>LRI (ligand-receptor interaction): a set of genes whose proteins are
known to interact during extracellular signalling. The package
distinguishes simple LRIs (e.g. <code>Apoe:Ldlr</code>) and
complex/heteromeric LRIs (e.g. <code>Col3a1:Itgb1_Itga2</code>).</li>
<li>CCI (cell-cell interaction): a communication signal of the form
<code>(B cell, T cell; Apoe:Ldlr)</code> where <code>B cell</code> is
the emitter cell type expressing the ligand <code>Apoe</code> and
<code>T cell</code> is the receiver cell type expressing the receptor
<code>Ldlr</code>.</li>
</ul>
<p><strong>Note: the toy-model results below do not convey any
meaningful biology.</strong></p>
</div>
<div class="section level2">
<h2 id="standard-intercellular-communication-differential-analysis">Standard intercellular communication differential analysis<a class="anchor" aria-label="anchor" href="#standard-intercellular-communication-differential-analysis"></a>
</h2>
<div class="section level3">
<h3 id="load-essential-libraries">Load essential libraries<a class="anchor" aria-label="anchor" href="#load-essential-libraries"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cyrillagger.github.io/scDiffCom/">scDiffCom</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prepare-a-seurat-object">Prepare a Seurat object<a class="anchor" aria-label="anchor" href="#prepare-a-seurat-object"></a>
</h3>
<p>The input of <code>scDiffCom</code> must be a <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> object with cells
annotated by cell types. A pairwise condition on the cells is also
necessary for the differential analysis.</p>
<p>As a toy-model example, we use a down-sampled Seurat object
containing mouse liver cells from Tabula Muris Senis. This object is
part of the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu">scDiffCom</span><span class="fu">::</span><span class="va"><a href="../reference/seurat_sample_tms_liver.html">seurat_sample_tms_liver</a></span></span>
<span><span class="va">seurat_object</span></span>
<span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 726 features across 468 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (726 features, 0 variable features)</span></span>
<span><span class="co">##  2 layers present: counts, data</span></span>
<span></span>
<span><span class="co"># The object already contains a column "cell_type" in its meta.data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">[[</span><span class="st">"cell_type"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## cell_type</span></span>
<span><span class="co">##                               B cell endothelial cell of hepatic sinusoid </span></span>
<span><span class="co">##                                   84                                  100 </span></span>
<span><span class="co">##                           hepatocyte                    myeloid leukocyte </span></span>
<span><span class="co">##                                  100                                  100 </span></span>
<span><span class="co">##                               T cell </span></span>
<span><span class="co">##                                   84</span></span>
<span></span>
<span><span class="co"># Cells can be grouped based on mice age</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">[[</span><span class="st">"age_group"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## age_group</span></span>
<span><span class="co">##   OLD YOUNG </span></span>
<span><span class="co">##   250   218</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-default-analysis">Run default analysis<a class="anchor" aria-label="anchor" href="#run-default-analysis"></a>
</h3>
<p>The full detection and differential analysis is performed by the
function <code>run_interaction_analysis</code>. Several permutation
tests are performed to assess the significance of both the specificity
and the differential expression of each potential cell-cell interaction
(CCI). 1000 permutations are done by default as this is fast and
sufficient for a preliminary exploratory analysis. Otherwise, we
recommend 10’000 permutations. Parallel computing can be easily enabled
by loading the <a href="https://cran.r-project.org/package=future" class="external-link">future</a> package and
setting the <code>plan</code> accordingly:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load future (optional)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">sequential</span><span class="op">)</span> <span class="co"># sequentially in the current R process, equivalent to do nothing</span></span>
<span><span class="co">#plan(multisession, workers = 4) # background R sessions</span></span>
<span><span class="co">#plan(multicore, workers = 4) # forked R processes, not Windows/not RStudio</span></span></code></pre></div>
<p>Running the analysis only requires to specify the species, the names
of the columns where the cell-type annotation and pairwise condition are
stored and the explicit names of the two conditions:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run differential analysis with default parameters</span></span>
<span><span class="va">scdiffcom_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_interaction_analysis.html">run_interaction_analysis</a></span><span class="op">(</span></span>
<span>  seurat_object <span class="op">=</span> <span class="va">seurat_object</span>,</span>
<span>  LRI_species <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>  seurat_celltype_id <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  seurat_condition_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    column_name <span class="op">=</span> <span class="st">"age_group"</span>,</span>
<span>    cond1_name <span class="op">=</span> <span class="st">"YOUNG"</span>,</span>
<span>    cond2_name <span class="op">=</span> <span class="st">"OLD"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Extracting data from assay 'RNA' and slot 'data' (assuming normalized log1p-transformed data).</span></span>
<span><span class="co">## Converting normalized data from log1p-transformed to non-log1p-transformed.</span></span>
<span><span class="co">## Input data: 726 genes, 468 cells and 5 cell-types.</span></span>
<span><span class="co">## Input ligand-receptor database: 4582 mouse interactions.</span></span>
<span><span class="co">## Number of LRIs that match to genes present in the dataset: 1173.</span></span>
<span><span class="co">## Type of analysis to be performed: differential analysis between YOUNG and OLD cells.</span></span>
<span><span class="co">## Total number of potential cell-cell interactions (CCIs): 29325 (5 * 5 * 1173).</span></span>
<span><span class="co">## Performing permutation analysis (1000 iterations by batches of 1000) on 8952 potential CCIs.</span></span>
<span><span class="co">## Performing batch 1 of 1.</span></span>
<span><span class="co">## Filtering and cleaning 'raw' CCIs.</span></span>
<span><span class="co">## Returning 1896 detected CCIs.</span></span>
<span><span class="co">## Performing over-representation analysis on the categories: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span></span>
<span><span class="co">## Successfully returning final scDiffCom object.</span></span></code></pre></div>
<p>The output of <code>run_interaction_analysis</code> is an S4 object
of class <code>scDiffCom</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scdiffcom_object</span></span>
<span><span class="co">## An object of class scDiffCom with name scDiffCom_object</span></span>
<span><span class="co">## Analysis performed: differential analysis between YOUNG and OLD cells</span></span>
<span><span class="co">## 1896 detected CCIs across 5 cell types</span></span>
<span><span class="co">## Over-representation results for LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="explore-the-results-from-the-shiny-app">Explore the results from the Shiny app<a class="anchor" aria-label="anchor" href="#explore-the-results-from-the-shiny-app"></a>
</h3>
<p>A Shiny app can be launched from each <code>scDiffCom</code> object
to interactively visualize the detected CCIs and how they are
differentially expressed. Over-representation results are also directly
available from the app. As an option, it is possible to first reduce
over-represented GO terms by semantic similarity such that they will be
displayed as a treemap. Also note that you might be prompted to install
several packages.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://shiny.posit.co/" class="external-link">"shiny"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"shiny"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/dreamRs/shinyWidgets" class="external-link">"shinyWidgets"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"shinyWidgets"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://rstudio.github.io/shinythemes/" class="external-link">"shinythemes"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"shinythemes"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/DT" class="external-link">"DT"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"DT"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">"kableExtra"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"kableExtra"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"RColorBrewer"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RColorBrewer"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://r.igraph.org/" class="external-link">"igraph"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"igraph"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://datastorm-open.github.io/visNetwork/" class="external-link">"visNetwork"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"visNetwork"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://plotly-r.com" class="external-link">"plotly"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"plotly"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">"GOSemSim"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"GoSemSim"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://www.bioconductor.org/packages/rrvgo" class="external-link">"rrvgo"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"rrvgo"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"org.Mm.eg.db"</span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"org.Mm.eg.db"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">reduced_go_terms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReduceGO.html">ReduceGO</a></span><span class="op">(</span><span class="va">scdiffcom_object</span><span class="op">)</span> <span class="co"># optional</span></span>
<span></span>
<span><span class="fu"><a href="../reference/BuildShiny.html">BuildShiny</a></span><span class="op">(</span></span>
<span>  <span class="va">scdiffcom_object</span>,</span>
<span>  reduced_go_table <span class="op">=</span> <span class="va">reduced_go_terms</span> <span class="co"># NULL by default</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="explore-the-results-manually">Explore the results manually<a class="anchor" aria-label="anchor" href="#explore-the-results-manually"></a>
</h3>
<p>Instead of using the Shiny app, the results can also be extracted
directly from the <code>scDiffCom</code> object with the provided
accessors:</p>
<ul>
<li>
<code><a href="../reference/GetParameters.html">GetParameters()</a></code> returns the list of parameters used by
<code><a href="../reference/run_interaction_analysis.html">run_interaction_analysis()</a></code>
</li>
<li>
<code><a href="../reference/GetTableCCI.html">GetTableCCI()</a></code> returns a (complete or simplified)
data.table either of the detected CCIs or of all the hypothetical
CCIs.</li>
<li>
<code><a href="../reference/GetTableORA.html">GetTableORA()</a></code> returns a list of (complete or
simplified) data.tables with over-representation results.</li>
</ul>
<p>From there, results can be analysed and visualised from either custom
or <code>scDiffCom</code> functions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Retrieve and display all detected CCIs</span></span>
<span><span class="va">CCI_detected</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetTableCCI.html">GetTableCCI</a></span><span class="op">(</span><span class="va">scdiffcom_object</span>, type <span class="op">=</span> <span class="st">"detected"</span>, simplified <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Number of CCIs per regulation type (here with age)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">CCI_detected</span><span class="op">$</span><span class="va">REGULATION</span><span class="op">)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## DOWN FLAT  NSC   UP </span></span>
<span><span class="co">##  115  927  820   34</span></span>
<span></span>
<span><span class="co"># Retrieve the ORA results</span></span>
<span><span class="va">ORA_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetTableORA.html">GetTableORA</a></span><span class="op">(</span><span class="va">scdiffcom_object</span>, categories <span class="op">=</span> <span class="st">"all"</span>, simplified <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Categories available</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ORA_results</span><span class="op">)</span></span>
<span><span class="co">## [1] "LRI"               "LIGAND_COMPLEX"    "RECEPTOR_COMPLEX" </span></span>
<span><span class="co">## [4] "ER_CELLTYPES"      "EMITTER_CELLTYPE"  "RECEIVER_CELLTYPE"</span></span>
<span><span class="co">## [7] "GO_TERMS"          "KEGG_PWS"</span></span></code></pre></div>
<p>Custom volcano plot of CCIs:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>  <span class="va">CCI_detected</span>,</span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>     x <span class="op">=</span> <span class="va">LOGFC</span>,</span>
<span>     y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">BH_P_VALUE_DE</span> <span class="op">+</span> <span class="fl">1E-2</span><span class="op">)</span>,</span>
<span>     colour <span class="op">=</span> <span class="va">REGULATION</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span></span>
<span>  values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UP"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"DOWN"</span> <span class="op">=</span> <span class="st">"blue"</span>, <span class="st">"FLAT"</span> <span class="op">=</span> <span class="st">"green"</span>, <span class="st">"NSC"</span> <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span></span>
<span>  <span class="st">"log(FC)"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span></span>
<span>  <span class="st">"-log10(Adj. p-value)"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiffCom-vignette_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
<p>Dot plots of over-represented terms:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the most over-represented up-regulated LRIs</span></span>
<span><span class="co"># PlotORA returns a ggplot object that you can further optimize (e.g. here to place the legend)</span></span>
<span><span class="fu"><a href="../reference/PlotORA.html">PlotORA</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">scdiffcom_object</span>,</span>
<span>  category <span class="op">=</span> <span class="st">"LRI"</span>,</span>
<span>  regulation <span class="op">=</span> <span class="st">"UP"</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>  legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.85</span>, <span class="fl">0.4</span><span class="op">)</span>,</span>
<span>  legend.key.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scDiffCom-vignette_files/figure-html/unnamed-chunk-9-1.png" width="100%"></p>
<p>A network of over-represented cell types and cell-type to cell-type
interactions:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://datastorm-open.github.io/visNetwork/" class="external-link">"visNetwork"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"visNetwork"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://r.igraph.org/" class="external-link">"igraph"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"igraph"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">"kableExtra"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"kableExtra"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"RColorBrewer"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RColorBrewer"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/BuildNetwork.html">BuildNetwork</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">scdiffcom_object</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div id="htmlwidget-409849e39179307f2e17" style="width:100%;height:100vh;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-409849e39179307f2e17">{"x":{"nodes":{"name":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","myeloid leukocyte (E)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"vertex_types":[true,true,true,true,true,false,false,false,false,false],"num_cells":[42,42,50,50,50,42,42,50,50,50],"ORA_TYPE":["UP","UP","NONE","FLAT","NONE","FLAT","DIFF","DOWN","NONE","FLAT"],"OR_UP":[5.26,3.03,0.381,0.2,0.574,0,30.5,0.0733,0,0.235],"OR_DOWN":[0.878,0.845,1.37,1.6,0.374,0.295,1.95,1.92,1.55,0.126],"OR_FLAT":[0.648,0.707,0.9330000000000001,1.52,1.09,1.66,0.507,0.893,1.33,1.33],"BH_P_VALUE_UP":[0.000402,0.0117,0.0498,0.0213,0.307,null,1.04e-16,0.000276,null,0.00697],"BH_P_VALUE_DOWN":[0.865,0.847,0.199,0.0759,0.00274,0.009129999999999999,0.0032,0.0032,0.121,4.57e-10],"BH_P_VALUE_FLAT":[0.0163,0.0163,0.487,0.000724,0.487,0.00231,1.88e-08,0.287,0.0699,0.00886],"vertex.size":[5,5,5,5,5,5,5,5,5,5],"id":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","myeloid leukocyte (E)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"label":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","myeloid leukocyte (E)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"value":[5,5,5,5,5,5,5,5,5,5],"color.background":["#F94144","#F94144","#a6a6a6","#000000","#a6a6a6","#000000","#F9C74F","#277DA1","#a6a6a6","#000000"],"color.border":["#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590"],"color.highlight.background":["#F94144","#F94144","#a6a6a6","#000000","#a6a6a6","#000000","#F9C74F","#277DA1","#a6a6a6","#000000"],"color.highlight.border":["#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590"],"color.hover.background":["#F94144","#F94144","#a6a6a6","#000000","#a6a6a6","#000000","#F9C74F","#277DA1","#a6a6a6","#000000"],"color.hover.border":["#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590","#577590"],"shadow":[true,true,true,true,true,true,true,true,true,true],"title":["<h4> B cell (E)<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio UP: <\/td>\n   <td style=\"text-align:left;\"> 5.26 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value UP: <\/td>\n   <td style=\"text-align:left;\"> 0.000402 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> T cell (E)<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio UP: <\/td>\n   <td style=\"text-align:left;\"> 3.03 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value UP: <\/td>\n   <td style=\"text-align:left;\"> 0.0117 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> endothelial cell of hepatic sinusoid (E)<\/h4>","<h4> hepatocyte (E)<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 1.52 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.000724 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> myeloid leukocyte (E)<\/h4>","<h4> B cell (R)<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 1.66 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.00231 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> T cell (R)<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio UP: <\/td>\n   <td style=\"text-align:left;\"> 30.5 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value UP: <\/td>\n   <td style=\"text-align:left;\"> 1.04e-16 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 1.95 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.0032 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> endothelial cell of hepatic sinusoid (R)<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 1.92 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.0032 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> hepatocyte (R)<\/h4>","<h4> myeloid leukocyte (R)<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 1.33 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.00886 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>"],"group":[true,true,true,true,true,false,false,false,false,false],"level":[1,1,1,1,1,2,2,2,2,2],"vertex_order":[1,2,4,5,3,7,6,9,10,8],"x":[-1,-1,-1,-1,-1,1,1,1,1,1],"y":[-1,-0.5,0.5,1,0,-0.5,-1,0.5,1,0]},"edges":{"from":["B cell (E)","T cell (E)","endothelial cell of hepatic sinusoid (E)","hepatocyte (E)","hepatocyte (E)","hepatocyte (E)","hepatocyte (E)","hepatocyte (E)"],"to":["T cell (R)","T cell (R)","hepatocyte (R)","B cell (R)","T cell (R)","endothelial cell of hepatic sinusoid (R)","hepatocyte (R)","myeloid leukocyte (R)"],"NUM_LRIS_TOTAL":[56,74,62,50,61,116,71,147],"NUM_LRIS_YOUNG":[38,57,59,44,59,113,71,145],"NUM_LRIS_OLD":[55,67,57,43,40,99,64,144],"NUM_LRIS_REL_DIFF":[17,10,-2,-1,-19,-14,-7,-1],"NUM_LRIS_UP":[10,9,0,0,2,0,0,0],"NUM_LRIS_DOWN":[1,5,10,1,14,15,5,2],"NUM_LRIS_FLAT":[16,28,33,35,23,54,48,93],"NUM_LRIS_NSC":[29,32,19,14,22,47,18,52],"NUM_LRIS_DIFF":[11,14,10,1,16,15,5,2],"ORA_TYPE":["UP","UP","DOWN","FLAT","DOWN","DOWN","FLAT","FLAT"],"OR_UP":[16.4,9.92,0,0,1.91,0,0,0],"OR_DOWN":[0.275,1.13,3.16,0.31,5.11,2.49,1.18,0.2],"OR_FLAT":[0.408,0.625,1.2,2.49,0.623,0.905,2.25,1.89],"BH_P_VALUE_UP":[1.52e-07,1.38e-05,null,null,0.539,null,null,null],"BH_P_VALUE_DOWN":[0.424,0.9429999999999999,0.0194,0.5600000000000001,0.000167,0.0194,0.821,0.0342],"BH_P_VALUE_FLAT":[0.0127,0.18,0.765,0.0127,0.25,0.82,0.0126,0.00736],"color.color":["#F94144","#F94144","#277DA1","#000000","#277DA1","#277DA1","#000000","#000000"],"main_title":["Cell-type centric over-representation","Cell-type centric over-representation","Cell-type centric over-representation","Cell-type centric over-representation","Cell-type centric over-representation","Cell-type centric over-representation","Cell-type centric over-representation","Cell-type centric over-representation"],"width":[3.039651145690063,3.855982944386633,3.324666526058841,2.738911960763933,3.278154143160369,5.4375685493849,3.727460773566809,6.425347053778234],"color.highlight":["#F94144","#F94144","#277DA1","#000000","#277DA1","#277DA1","#000000","#000000"],"color.hover":["#F94144","#F94144","#277DA1","#000000","#277DA1","#277DA1","#000000","#000000"],"smooth":[true,true,true,true,true,true,true,true],"arrow.size":[0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5],"title":["<h4> B cell to T cell<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio UP: <\/td>\n   <td style=\"text-align:left;\"> 16.4 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value UP: <\/td>\n   <td style=\"text-align:left;\"> 1.52e-07 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 56 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 10 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 1 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 16 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> T cell to T cell<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio UP: <\/td>\n   <td style=\"text-align:left;\"> 9.92 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value UP: <\/td>\n   <td style=\"text-align:left;\"> 1.38e-05 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 74 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 9 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 5 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 28 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> endothelial cell of hepatic sinusoid to hepatocyte<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 3.16 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.0194 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 62 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 0 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 10 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 33 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> hepatocyte to B cell<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 2.49 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.0127 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 50 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 0 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 1 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 35 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> hepatocyte to T cell<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 5.11 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.000167 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 61 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 2 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 14 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 23 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> hepatocyte to endothelial cell of hepatic sinusoid<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio DOWN: <\/td>\n   <td style=\"text-align:left;\"> 2.49 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value DOWN: <\/td>\n   <td style=\"text-align:left;\"> 0.0194 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 116 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 0 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 15 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 54 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> hepatocyte to hepatocyte<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 2.25 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.0126 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 71 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 0 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 5 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 48 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>","<h4> hepatocyte to myeloid leukocyte<\/h4><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> Odds Ratio FLAT: <\/td>\n   <td style=\"text-align:left;\"> 1.89 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> Adj. p-value FLAT: <\/td>\n   <td style=\"text-align:left;\"> 0.00736 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table><br> Number of LRIs: <br><table>\n<tbody>\n  <tr>\n   <td style=\"text-align:left;\"> TOTAL: <\/td>\n   <td style=\"text-align:left;\"> 147 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> UP: <\/td>\n   <td style=\"text-align:left;\"> 0 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> DOWN: <\/td>\n   <td style=\"text-align:left;\"> 2 <\/td>\n  <\/tr>\n  <tr>\n   <td style=\"text-align:left;\"> FLAT: <\/td>\n   <td style=\"text-align:left;\"> 93 <\/td>\n  <\/tr>\n<\/tbody>\n<\/table>"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","font":{"size":18,"align":"left"},"physics":false},"manipulation":{"enabled":false},"edges":{"arrows":"middle","shadow":true,"smooth":false},"physics":{"maxVelocity":10,"timestep":0.5,"enabled":false,"repulsion":{"damping":0.5}},"interaction":{"hover":true,"keyboard":{"enabled":true},"multiselect":true,"navigationButtons":false,"zoomSpeed":1},"autoResize":true},"groups":["TRUE","FALSE"],"width":"100%","height":"100vh","idselection":{"enabled":false,"style":"width: 150px; height: 26px","useLabels":true,"main":"Select by id"},"byselection":{"enabled":false,"style":"width: 150px; height: 26px","multiple":false,"hideColor":"rgba(200,200,200,0.5)","highlight":false},"main":{"text":"Cell-type centric over-representation","style":"font-family:Georgia, Times New Roman, Times, serif;font-weight:bold;font-size:20px;text-align:center;"},"submain":{"text":"YOUNG vs OLD","style":"font-family:Georgia, Times New Roman, Times, serif;font-size:12px;text-align:center;"},"footer":null,"background":"#F5F5F5","igraphlayout":{"type":"square"},"legend":{"width":0.2,"useGroups":false,"position":"left","ncol":1,"stepX":100,"stepY":100,"zoom":false,"edges":{"color":["#F94144","#277DA1","#000000","#F9C74F"],"label":["UP","DOWN","FLAT","UP&DOWN"],"arrows":["to","to","to","to"],"font.align":["top","top","top","top"]},"edgesToDataframe":true,"nodes":{"color":["#F94144","#277DA1","#000000","#F9C74F"],"label":["UP","DOWN","FLAT","UP&DOWN"],"shape":["dot","dot","dot","dot"]},"nodesToDataframe":true},"opts_manipulation":{"datacss":"table.legend_table {\n  font-size: 11px;\n  border-width:1px;\n  border-color:#d3d3d3;\n  border-style:solid;\n}\ntable.legend_table td {\n  border-width:1px;\n  border-color:#d3d3d3;\n  border-style:solid;\n  padding: 2px;\n}\ndiv.table_content {\n  width:80px;\n  text-align:center;\n}\ndiv.table_description {\n  width:100px;\n}\n\n.operation {\n  font-size:20px;\n}\n\n.network-popUp {\n  display:none;\n  z-index:299;\n  width:250px;\n  /*height:150px;*/\n  background-color: #f9f9f9;\n  border-style:solid;\n  border-width:1px;\n  border-color: #0d0d0d;\n  padding:10px;\n  text-align: center;\n  position:fixed;\n  top:50%;  \n  left:50%;  \n  margin:-100px 0 0 -100px;  \n\n}","addNodeCols":["id","label"],"editNodeCols":["id","label"],"tab_add_node":"<span id=\"addnode-operation\" class = \"operation\">node<\/span> <br><table style=\"margin:auto;\"><tr><td>id<\/td><td><input id=\"addnode-id\"  type= \"text\" value=\"new value\"><\/td><\/tr><tr><td>label<\/td><td><input id=\"addnode-label\"  type= \"text\" value=\"new value\"><\/td><\/tr><\/table><input type=\"button\" value=\"save\" id=\"addnode-saveButton\"><\/button><input type=\"button\" value=\"cancel\" id=\"addnode-cancelButton\"><\/button>","tab_edit_node":"<span id=\"editnode-operation\" class = \"operation\">node<\/span> <br><table style=\"margin:auto;\"><tr><td>id<\/td><td><input id=\"editnode-id\"  type= \"text\" value=\"new value\"><\/td><\/tr><tr><td>label<\/td><td><input id=\"editnode-label\"  type= \"text\" value=\"new value\"><\/td><\/tr><\/table><input type=\"button\" value=\"save\" id=\"editnode-saveButton\"><\/button><input type=\"button\" value=\"cancel\" id=\"editnode-cancelButton\"><\/button>"},"highlight":{"enabled":true,"hoverNearest":true,"degree":{"from":1,"to":1},"algorithm":"hierarchical","hideColor":"rgba(200,200,200,0.5)","labelOnly":true},"collapse":{"enabled":false,"fit":false,"resetHighlight":true,"clusterOptions":null,"keepCoord":true,"labelSuffix":"(cluster)"},"tooltipStay":300,"tooltipStyle":"position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #f5f4ed;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div class="section level2">
<h2 id="detection-only-analysis">Detection-only analysis<a class="anchor" aria-label="anchor" href="#detection-only-analysis"></a>
</h2>
<p>If one has no condition to compare, <code>scDiffCom</code> can still
be used to return detected CCIs only:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scdiffcom_detection_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_interaction_analysis.html">run_interaction_analysis</a></span><span class="op">(</span></span>
<span>  seurat_object <span class="op">=</span> <span class="va">seurat_object</span>,</span>
<span>  LRI_species <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>  seurat_celltype_id <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  seurat_condition_id <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Extracting data from assay 'RNA' and slot 'data' (assuming normalized log1p-transformed data).</span></span>
<span><span class="co">## Converting normalized data from log1p-transformed to non-log1p-transformed.</span></span>
<span><span class="co">## Input data: 726 genes, 468 cells and 5 cell-types.</span></span>
<span><span class="co">## Input ligand-receptor database: 4582 mouse interactions.</span></span>
<span><span class="co">## Number of LRIs that match to genes present in the dataset: 1173.</span></span>
<span><span class="co">## Type of analysis to be performed: detection analysis without conditions.</span></span>
<span><span class="co">## Total number of potential cell-cell interactions (CCIs): 29325 (5 * 5 * 1173).</span></span>
<span><span class="co">## Performing permutation analysis (1000 iterations by batches of 1000) on 7352 potential CCIs.</span></span>
<span><span class="co">## Performing batch 1 of 1.</span></span>
<span><span class="co">## Filtering and cleaning 'raw' CCIs.</span></span>
<span><span class="co">## Returning 1947 detected CCIs.</span></span>
<span><span class="co">## No over-representation analysis available for the selected parameters.</span></span>
<span><span class="co">## Successfully returning final scDiffCom object.</span></span></code></pre></div>
<p>Although it is currently not possible to explore such results with a
Shiny app, such functionality should be added in a future release of the
package ( as well as with other downstream analysis tools).</p>
</div>
<div class="section level2">
<h2 id="database-of-ligand-receptor-interactions">Database of ligand-receptor interactions<a class="anchor" aria-label="anchor" href="#database-of-ligand-receptor-interactions"></a>
</h2>
<p><code>scDiffcom</code> infers cell type to cell type communication
patterns from the expression of genes known to be involved in
ligand-receptor interactions (LRIs). The package contains it own
internal databases of curated LRIs (for human, mouse and rat), retrieved
from previous studies.</p>
<p>As shown above, there is no need to call the LRI databases explicitly
when performing an analysis. However, they can be accessed and explored
as follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load, e.g., the mouse database</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">LRI_mouse</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the data.table of LRIs (more information available in other columns)</span></span>
<span><span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LRI"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##                        LRI</span></span>
<span><span class="co">##    1: 1700013F07Rik:Plscr4</span></span>
<span><span class="co">##    2:  9530003J23Rik:Itgal</span></span>
<span><span class="co">##    3:             A2m:Lrp1</span></span>
<span><span class="co">##    4:         Aanat:Mtnr1a</span></span>
<span><span class="co">##    5:         Aanat:Mtnr1b</span></span>
<span><span class="co">##   ---                     </span></span>
<span><span class="co">## 4578:               a:Mc1r</span></span>
<span><span class="co">## 4579:               a:Mc2r</span></span>
<span><span class="co">## 4580:               a:Mc3r</span></span>
<span><span class="co">## 4581:               a:Mc4r</span></span>
<span><span class="co">## 4582:               a:Mc5r</span></span>
<span></span>
<span><span class="co"># Display the data.table of GO Terms attached to each LRI</span></span>
<span><span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated_GO</span></span>
<span><span class="co">##                          LRI      GO_ID</span></span>
<span><span class="co">##      1: 1700013F07Rik:Plscr4 GO:0005575</span></span>
<span><span class="co">##      2: 1700013F07Rik:Plscr4 GO:0110165</span></span>
<span><span class="co">##      3: 1700013F07Rik:Plscr4 GO:0003674</span></span>
<span><span class="co">##      4: 1700013F07Rik:Plscr4 GO:0005215</span></span>
<span><span class="co">##      5: 1700013F07Rik:Plscr4 GO:0005488</span></span>
<span><span class="co">##     ---                                </span></span>
<span><span class="co">## 247958:               a:Mc5r GO:0065007</span></span>
<span><span class="co">## 247959:               a:Mc5r GO:0050789</span></span>
<span><span class="co">## 247960:               a:Mc5r GO:0050794</span></span>
<span><span class="co">## 247961:               a:Mc5r GO:0007165</span></span>
<span><span class="co">## 247962:               a:Mc5r GO:0019222</span></span>
<span></span>
<span><span class="co"># Information about GO terms (more information available in other columns)</span></span>
<span><span class="co"># LEVEL corresponds to the depth of the GO term in the GO graph</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gene_ontology_level</span><span class="op">)</span></span>
<span><span class="va">gene_ontology_level</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NAME"</span>, <span class="st">"LEVEL"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##                                                         NAME LEVEL</span></span>
<span><span class="co">##     1:                             mitochondrion inheritance     8</span></span>
<span><span class="co">##     2:                      mitochondrial genome maintenance     7</span></span>
<span><span class="co">##     3:                                          reproduction     2</span></span>
<span><span class="co">##     4:                 obsolete ribosomal chaperone activity     1</span></span>
<span><span class="co">##     5: high-affinity zinc transmembrane transporter activity    10</span></span>
<span><span class="co">##    ---                                                            </span></span>
<span><span class="co">## 47216:                                               part of     1</span></span>
<span><span class="co">## 47217:                                  positively regulates     2</span></span>
<span><span class="co">## 47218:                                             regulates     1</span></span>
<span><span class="co">## 47219:                                         starts_during     1</span></span>
<span><span class="co">## 47220:                                     term tracker item     1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="using-a-custom-database-of-ligand-receptor-interactions">Using a custom database of ligand-receptor interactions<a class="anchor" aria-label="anchor" href="#using-a-custom-database-of-ligand-receptor-interactions"></a>
</h2>
<p><code>scDiffcom</code> also allows users to use their own LRI
database and associated annotations as follows.</p>
<p>As an example, we simply use a downsized version of the mouse
database. It can be any database as long as it is a data.table with the
same columns as the default.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_LRI</span> <span class="op">&lt;-</span> <span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="op">]</span></span></code></pre></div>
<p>We note that ORA on GO and KEGG terms cannot be peformed by default
when using a custom database. It is therefore skipped unless the user
provides custom annotation tables. Here, we can obtain them by
subsetting the default tables accordingly:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_LRI_GO</span> <span class="op">&lt;-</span> <span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated_GO</span><span class="op">[</span><span class="va">LRI</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">custom_LRI</span><span class="op">$</span><span class="va">LRI</span><span class="op">]</span></span>
<span><span class="va">custom_LRI_KEGG</span> <span class="op">&lt;-</span> <span class="va">LRI_mouse</span><span class="op">$</span><span class="va">LRI_curated_KEGG</span><span class="op">[</span><span class="va">LRI</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">custom_LRI</span><span class="op">$</span><span class="va">LRI</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LRI"</span>, <span class="st">"KEGG_ID"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Finally, scDiffCom can be run by changing the parameter
<code>LRI_species</code> to <code>"custom"</code> and providing the
custom LRI and annotation tables as a list:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scdiffcom_object_customlri</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/run_interaction_analysis.html">run_interaction_analysis</a></span><span class="op">(</span></span>
<span>  seurat_object <span class="op">=</span> <span class="va">seurat_object</span>,</span>
<span>  LRI_species <span class="op">=</span> <span class="st">"custom"</span>,</span>
<span>  seurat_celltype_id <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  seurat_condition_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    column_name <span class="op">=</span> <span class="st">"age_group"</span>,</span>
<span>    cond1_name <span class="op">=</span> <span class="st">"YOUNG"</span>,</span>
<span>    cond2_name <span class="op">=</span> <span class="st">"OLD"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  custom_LRI_tables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>LRI <span class="op">=</span> <span class="va">custom_LRI</span>, custom_GO <span class="op">=</span> <span class="va">custom_LRI_GO</span>, custom_KEGG <span class="op">=</span> <span class="va">custom_LRI_KEGG</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Using custom LRI table. Use at your own risk!</span></span>
<span><span class="co">## Extracting data from assay 'RNA' and slot 'data' (assuming normalized log1p-transformed data).</span></span>
<span><span class="co">## Converting normalized data from log1p-transformed to non-log1p-transformed.</span></span>
<span><span class="co">## Input data: 726 genes, 468 cells and 5 cell-types.</span></span>
<span><span class="co">## Input ligand-receptor database: 100 custom interactions.</span></span>
<span><span class="co">## Number of LRIs that match to genes present in the dataset: 16.</span></span>
<span><span class="co">## Type of analysis to be performed: differential analysis between YOUNG and OLD cells.</span></span>
<span><span class="co">## Total number of potential cell-cell interactions (CCIs): 400 (5 * 5 * 16).</span></span>
<span><span class="co">## Performing permutation analysis (1000 iterations by batches of 1000) on 111 potential CCIs.</span></span>
<span><span class="co">## Performing batch 1 of 1.</span></span>
<span><span class="co">## Filtering and cleaning 'raw' CCIs.</span></span>
<span><span class="co">## Returning 18 detected CCIs.</span></span>
<span><span class="co">## ORA not performed on GO/KEGG when using a custom LRI database, unless custom tables are provided.</span></span>
<span><span class="co">## Performing over-representation analysis on the categories: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_ID, KEGG_ID.</span></span>
<span><span class="co">## Successfully returning final scDiffCom object.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="additional-functionalities">Additional functionalities<a class="anchor" aria-label="anchor" href="#additional-functionalities"></a>
</h2>
<div class="section level3">
<h3 id="modifying-default-parameters">Modifying default parameters<a class="anchor" aria-label="anchor" href="#modifying-default-parameters"></a>
</h3>
<p><code>run_interaction_analysis</code> depends on several parameters
that can be modified. We encourage to carefully read the documentation
(<code><a href="../reference/run_interaction_analysis.html">?run_interaction_analysis</a></code>). For convenience, the input
parameters are stored in the returned object in the slot
<code>parameters</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># display the first three parameters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/GetParameters.html">GetParameters</a></span><span class="op">(</span><span class="va">scdiffcom_object</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">## $object_name</span></span>
<span><span class="co">## [1] "scDiffCom_object"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $LRI_species</span></span>
<span><span class="co">## [1] "mouse"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $seurat_celltype_id</span></span>
<span><span class="co">## [1] "cell_type"</span></span></code></pre></div>
<p>Among those parameters are four filtering thresholds that can be
modified to recompute the results without having to perform the
time-consuming permutation analyses again. This works because we store
an intermediate table of all hypothetic CCIs in the slot
<code>cci_table_raw</code>. The disadvantage is that storing this
data.table can make the object heavy depending on the dataset. You can
either decide to keep this table allowing you to quickly update your
results in case you want to modify some filtering parameters or to erase
this table to save space.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a new object with a different log fold change threshold</span></span>
<span><span class="va">scdiffcom_object_stringent_logfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterCCI.html">FilterCCI</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">scdiffcom_object</span>,</span>
<span>  new_threshold_logfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Filtering and cleaning 'raw' CCIs.</span></span>
<span><span class="co">## Returning 1896 detected CCIs.</span></span>
<span><span class="co">## Performing over-representation analysis on the categories: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span></span>
<span><span class="co">## Erasing all previous ORA results: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span></span>
<span></span>
<span><span class="co"># create a copy of the original object, but without the heavy intermediate table</span></span>
<span><span class="va">scdiffcom_object_light</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EraseRawCCI.html">EraseRawCCI</a></span><span class="op">(</span><span class="va">scdiffcom_object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the light object cannot be updated anymore</span></span>
<span><span class="fu"><a href="../reference/FilterCCI.html">FilterCCI</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">scdiffcom_object_light</span>,</span>
<span>  new_threshold_logfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Error in run_filtering_and_ora(object = object, new_threshold_quantile_score = new_threshold_quantile_score, : slot 'cci_table_raw' of 'object' is empty, impossible to filter</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="performing-ora-on-custom-categories">Performing ORA on custom categories<a class="anchor" aria-label="anchor" href="#performing-ora-on-custom-categories"></a>
</h3>
<p>By default, ORA is performed on a set of categories that are always
present for any dataset analysed (LRI, GO Terms, etc). However, you
might want to find over-represented results for a category only relevant
to your dataset. This is possible to achieve as long as you are able to
build a relationship between the terms of your new category and those of
a default category. As an example, we show here how we can perform ORA
on cell-type families:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a data.table that associates cell-type families to cell types</span></span>
<span><span class="co"># Here we use the default category EMITTER_CELLTYPE (RECEIVER_CELLTYPE would also work)</span></span>
<span><span class="va">cell_families_relation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></span><span class="op">(</span></span>
<span>  EMITTER_CELLTYPE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"B cell"</span>,</span>
<span>    <span class="st">"T cell"</span>,</span>
<span>    <span class="st">"endothelial cell of hepatic sinusoid"</span>,</span>
<span>    <span class="st">"hepatocyte"</span>,</span>
<span>    <span class="st">"myeloid leukocyte"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  EMITTER_CELLFAMILY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"leukocyte"</span>,</span>
<span>    <span class="st">"leukocyte"</span>,</span>
<span>    <span class="st">"endothelial cell"</span>,</span>
<span>    <span class="st">"epithelial cell"</span>,</span>
<span>    <span class="st">"leukocyte"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run ORA with the cell type families as extra annotation</span></span>
<span><span class="va">scdiffcom_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunORA.html">RunORA</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">scdiffcom_object</span>,</span>
<span>  extra_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="va">cell_families_relation</span></span>
<span>  <span class="op">)</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Performing over-representation analysis on the categories: EMITTER_CELLFAMILY.</span></span>
<span><span class="co">## Keeping previous ORA results: LRI, LIGAND_COMPLEX, RECEPTOR_COMPLEX, ER_CELLTYPES, EMITTER_CELLTYPE, RECEIVER_CELLTYPE, GO_TERMS, KEGG_PWS.</span></span></code></pre></div>
<p>The object now contains a new table in its slot
<code>ora_table</code>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Cyril Lagger, Eugen Ursu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
